src/snowflake/snowpark/profiler.py

```diff
@@ -0,0 +1,125 @@
+#
+from contextlib import contextmanager
+from typing import List, Optional
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.utils import validate_object_name
+
+
+class Profiler:
+    def __init__(
+        self,
+        stage: Optional[str] = "",
+        active_profiler: Optional[str] = "LINE",
+        session: Optional["snowflake.snowpark.Session"] = None,
+    ) -> None:
+        self.stage = stage
+        self.active_profiler = active_profiler
+        self.modules_to_register = []
+        self.register_modules_sql = ""
+        self.set_targeted_stage_sql = ""
+        self.enable_profiler_sql = ""
+        self.disable_profiler_sql = ""
+        self.set_active_profiler_sql = ""
+        self.session = session
+        self.prepare_sql()
+        self.query_history = None
+
+    def prepare_sql(self):
+        self.register_modules_sql = f"alter session set python_profiler_modules='{','.join(self.modules_to_register)}'"
+        self.set_targeted_stage_sql = (
+            f'alter session set PYTHON_PROFILER_TARGET_STAGE ="{self.stage}"'
+        )
+        self.enable_profiler_sql = "alter session set ENABLE_PYTHON_PROFILER = true"
+        self.disable_profiler_sql = "alter session set ENABLE_PYTHON_PROFILER = false"
+        self.set_active_profiler_sql = f"alter session set ACTIVE_PYTHON_PROFILER = '{self.active_profiler.upper()}'"
+
+    def register_profiler_modules(self, modules: List[str]):
+        self.modules_to_register = modules
+        self.prepare_sql()
+        if self.session is not None:
+            self._register_modules()
+
+    def set_targeted_stage(self, stage: str):
+        validate_object_name(stage)
+        self.stage = stage
+        self.prepare_sql()
+        if self.session is not None:
+            self._set_targeted_stage()
+
+    def set_active_profiler(self, active_profiler: str):
+        if self.active_profiler not in ["LINE", "MEMORY"]:
+            raise ValueError(
+                f"active_profiler expect 'LINE' or 'MEMORY', got {self.active_profiler} instead"
+            )
+        self.active_profiler = active_profiler
+        self.prepare_sql()
+        if self.session is not None:
+            self._set_active_profiler()
+
+    def _register_modules(self):
+        self.session.sql(self.register_modules_sql).collect()
+
+    def _set_targeted_stage(self):
+        self.session.sql(self.set_targeted_stage_sql).collect()
+
+    def _set_active_profiler(self):
+        self.session.sql(self.set_active_profiler_sql).collect()
+
+    def enable_profiler(self):
+        self.session.sql(self.enable_profiler_sql).collect()
+
+    def disable_profiler(self):
+        self.session.sql(self.disable_profiler_sql).collect()
+
+    def _get_last_query_id(self):
+        sps = self.session.sql("show procedures").collect()
+        names = [r.name for r in sps]
+        for query in self.query_history.queries[::-1]:
+            if query.sql_text.startswith("CALL"):
```

Comment:
there are some corner case like anonymous stored proc, we can probably loosen the requirement a bit.

Alternatively, we can just list all the files on the profiling stage, and print out the last modified file, that might be more robust than relying on query id.

Reply:
That works, one potential problem is if multiple session set same targeted profiler stage, the result will be inaccurate. We would need to ask user avoid using same stage if use this method. But generally I think this is more robust.

Reply:
Synced offline, let's relax the string matching with regex to accommodate anonymous sproc. https://docs.snowflake.com/en/sql-reference/sql/call-with